testcases/kernel/device-drivers/uart/serialcheck.sh

```diff
@@ -0,0 +1,95 @@
+#! /bin/sh
+###############################################################################
+# Copyright (C) 2015 Texas Instruments Incorporated - http://www.ti.com/
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation version 2.
+#
+# This program is distributed "as is" WITHOUT ANY WARRANTY of any
+# kind, whether express or implied; without even the implied warranty
+# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+###############################################################################
+
+source "common.sh"  # Import do_cmd(), die() and other functions
+
+#### Functions definitions ####
+usage()
+{
+    echo "usage: ./${0##*/} [-r UART_RATE] [-l LOOPS] [-x to enable HW flow control]"
+    exit 1
+}
+
+check_requirements() {
+    which serialcheck || die "serialcheck binary is not available. Please install it"
+}
+
+create_test_file() {
+    temp_test_file=`mktemp`
+    dd if=/dev/urandom of=$temp_test_file count=1 bs=$((UART_RATE / 2))
+}
+
+get_uart_ports() {
+for i in ${ARRAY[@]}; do
+    PORT=/dev/`echo $i | cut -d'/' -f 5`
+    # Activate port in case it will be initialized only when startup
+    echo "DDT TESTING" > $PORT 2>/dev/null
+    if [ `cat $i` -ne 0 ]; then
+        UART_PORTS=("${UART_PORTS[@]}" "$PORT")
+    fi
+done
+}
+
+filter_out_used_ports() {
+    for i in ${UART_PORTS[@]}; do
+        lsof | grep $i &> /dev/null || PORTS_TO_TEST=("${PORTS_TO_TEST[@]}" $i)
+    done
+}
+
+run_serial_check() {
+    for i in ${PORTS_TO_TEST[@]}; do
+        if [ $UART_HWFLOW -eq 0 ]; then
+            echo ''; echo "Testing $i in loopback at $UART_RATE $UART_LOOPS times"
+            { sleep 1; serialcheck -b $UART_RATE -d $i -f $temp_test_file -l $UART_LOOPS -m t -k; }&
+            PID=$!
+            serialcheck -b $UART_RATE -d $i -f $temp_test_file -l $UART_LOOPS -m r -k || { kill -- -$PID 2>/dev/null; die "TEST FAILED"; }
+        else
+            echo ''; echo "Testing $i with HW flow control at $UART_RATE $UART_LOOPS times"
+            { sleep 1; serialcheck -b $UART_RATE -d $i -f $temp_test_file -l $UART_LOOPS -m t -h; } &
+            PID=$!
+            serialcheck -b $UART_RATE -d $i -f $temp_test_file -l $UART_LOOPS -m r -h || { kill -- -$PID 2>/dev/null; die "TEST FAILED"; }
+        fi
+    done
+    rm $temp_test_file
```

Comment:
And this actually requires loopback to be set up on the port right?

Shouldn't we pass the information about ports to test to the test instead of trying all ports on the system?


I really do wonder what is a proper way to run such test. I guess that the information about port(s) that have tx and rx connected has to be passed to the testruner somehow so that we don't have to discover them here.

Reply:
It could be possible to create a static file that maps platforms to serial ports to test but we are trying to avoid that to save maintenance. That would force us to modify the file every time a new platform is added.